Typescript compilation fails on 4.0.0, Cannot find name 'this'

**Describe the bug**

Typescript fail to compile, that is a simple error that can be fixed in a patch

node_modules/whatsapp-api-js/lib/types/index.d.ts:155:139 - error TS2304: Cannot find name 'this'.

155     broadcastMessage(phoneID: string, to: string[], message: ClientMessage, batch_size?: number, delay?: number): Array<ReturnType<typeof this.sendMessage>>;
                                            

**Expected behavior**
tsc should pass...

**Desktop (please complete the following information):**

-   OS: Windows
-   Runtime node@20
-   Library Version 4.0.0

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
     "name": "whatsapp-api-js",
-    "version": "4.0.0",
+    "version": "4.0.1",
     "author": "Secreto31126",
     "description": "A TypeScript server agnostic Whatsapp's Official API framework",
     "license": "MIT",

--- a/src/index.ts
+++ b/src/index.ts
@@ -322,8 +322,8 @@ export class WhatsAppAPI<EmittersReturnType = void> {
         message: ClientMessage,
         batch_size = 50,
         delay = 1000
-    ): Array<ReturnType<typeof this.sendMessage>> {
-        const responses = [] as ReturnType<typeof this.sendMessage>[];
+    ): Array<ReturnType<WhatsAppAPI["sendMessage"]>> {
+        const responses = [] as ReturnType<WhatsAppAPI["sendMessage"]>[];
 
         if (batch_size < 1) {
             throw new RangeError("batch_size must be greater than 0");

--- a/src/messages/contacts.ts
+++ b/src/messages/contacts.ts
@@ -126,7 +126,7 @@ export class Contacts extends ClientMessage {
      * @override
      * @internal
      */
-    _build() {
+    _build(): string {
         return JSON.stringify(this.component);
     }
 }
@@ -245,7 +245,7 @@ export class Birthday extends ContactUniqueComponent {
      * @override
      * @internal
      */
-    _build() {
+    _build(): string {
         return this.birthday;
     }
 }

--- a/src/messages/template.ts
+++ b/src/messages/template.ts
@@ -157,7 +157,11 @@ export class Template extends ClientMessage {
      * @param code - The one time password to be sent
      * @returns A Template object for the API
      */
-    static OTP(name: string, language: string | Language, code: string) {
+    static OTP(
+        name: string,
+        language: string | Language,
+        code: string
+    ): Template {
         return new Template(
             name,
             language,
@@ -324,7 +328,7 @@ export abstract class ButtonComponent implements TemplateComponent {
      * @override
      * @internal
      */
-    _build(pointers: BuildingPointers) {
+    _build(pointers: BuildingPointers): this {
         this.index = pointers.button_counter++;
         return this;
     }
@@ -603,7 +607,7 @@ export class SkipButtonComponent extends ButtonComponent {
      * @override
      * @internal
      */
-    _build(pointers: BuildingPointers) {
+    _build(pointers: BuildingPointers): this {
         pointers.button_counter++;
         return null as unknown as this;
     }
@@ -638,7 +642,7 @@ export class HeaderComponent implements TemplateComponent {
      * @override
      * @internal
      */
-    _build() {
+    _build(): this {
         return this;
     }
 }
@@ -762,7 +766,7 @@ export class BodyComponent implements TemplateComponent {
      * @internal
      * @throws If theres_only_body is false and one of the parameters is a string and it's over 1024 characters
      */
-    _build({ theres_only_body }: BuildingPointers) {
+    _build({ theres_only_body }: BuildingPointers): this {
         // If it needs to check for the shorter max text length
         if (!theres_only_body) {
             for (const param of this.parameters) {
@@ -868,7 +872,7 @@ export class CarouselComponent
      * @override
      * @internal
      */
-    _build() {
+    _build(): this {
         return this;
     }
 }
@@ -919,7 +923,7 @@ export class CarouselCard implements ClientBuildableMessageComponent {
      * @override
      * @internal
      */
-    _build(ptr: { counter: number }) {
+    _build(ptr: { counter: number }): this {
         this.card_index = ptr.counter++;
         return this;
     }
@@ -974,7 +978,7 @@ export class LTOComponent implements TemplateComponent {
      * @override
      * @internal
      */
-    _build() {
+    _build(): this {
         return this;
     }
 }

--- a/src/middleware/adonis.ts
+++ b/src/middleware/adonis.ts
@@ -31,7 +31,7 @@ export class WhatsAppAPI extends WhatsAppAPIMiddleware {
      * @param req - The request object from AdonisJS
      * @returns The status code to be sent to the client
      */
-    async handle_post(req: Request) {
+    async handle_post(req: Request): Promise<number> {
         try {
             await this.post(
                 req.body() as PostData,
@@ -74,7 +74,7 @@ export class WhatsAppAPI extends WhatsAppAPIMiddleware {
      * @returns The challenge string to be sent to the client
      * @throws The error code
      */
-    handle_get(req: Request) {
+    handle_get(req: Request): string {
         try {
             return this.get(req.qs() as GetParams);
         } catch (e) {

--- a/src/middleware/bun.ts
+++ b/src/middleware/bun.ts
@@ -29,7 +29,7 @@ export class WhatsAppAPI extends WebStandardMiddleware {
      * @param req - The request object
      * @returns The status code to be sent to the client
      */
-    async handle_post(req: Request) {
+    async handle_post(req: Request): Promise<number> {
         return super.handle_post(req);
     }
 
@@ -61,7 +61,7 @@ export class WhatsAppAPI extends WebStandardMiddleware {
      * @returns The challenge string to be sent to the client
      * @throws The error code
      */
-    handle_get(req: Request) {
+    handle_get(req: Request): string {
         return super.handle_get(req);
     }
 }

--- a/src/middleware/deno.ts
+++ b/src/middleware/deno.ts
@@ -29,7 +29,7 @@ export class WhatsAppAPI extends WebStandardMiddleware {
      * @param req - The request object
      * @returns The status code to be sent to the client
      */
-    async handle_post(req: Request) {
+    async handle_post(req: Request): Promise<number> {
         return super.handle_post(req);
     }
 
@@ -61,7 +61,7 @@ export class WhatsAppAPI extends WebStandardMiddleware {
      * @returns The challenge string to be sent to the client
      * @throws The error code
      */
-    handle_get(req: Request) {
+    handle_get(req: Request): string {
         return super.handle_get(req);
     }
 }

--- a/src/middleware/express.ts
+++ b/src/middleware/express.ts
@@ -41,7 +41,7 @@ export class WhatsAppAPI extends WhatsAppAPIMiddleware {
      * @param req - The request object from Express.js
      * @returns The status code to be sent to the client
      */
-    async handle_post(req: Request) {
+    async handle_post(req: Request): Promise<number> {
         try {
             await this.post(
                 JSON.parse(req.body ?? "{}"),
@@ -85,7 +85,7 @@ export class WhatsAppAPI extends WhatsAppAPIMiddleware {
      * @returns The challenge string to be sent to the client
      * @throws The error code
      */
-    handle_get(req: Request) {
+    handle_get(req: Request): string {
         try {
             return this.get(req.query as GetParams);
         } catch (e) {

--- a/src/middleware/next.ts
+++ b/src/middleware/next.ts
@@ -27,7 +27,7 @@ export class WhatsAppAPI extends WebStandardMiddleware {
      * @param req - The request object
      * @returns The status code to be sent to the client
      */
-    async handle_post(req: Request) {
+    async handle_post(req: Request): Promise<number> {
         return super.handle_post(req);
     }
 
@@ -57,7 +57,7 @@ export class WhatsAppAPI extends WebStandardMiddleware {
      * @returns The challenge string to be sent to the client
      * @throws The error code
      */
-    handle_get(req: Request) {
+    handle_get(req: Request): string {
         return super.handle_get(req);
     }
 }

--- a/src/middleware/node-http.ts
+++ b/src/middleware/node-http.ts
@@ -38,7 +38,7 @@ export class WhatsAppAPI extends WhatsAppAPIMiddleware {
      * @param req - The request object
      * @returns The status code to be sent to the client
      */
-    async handle_post(req: IncomingMessage) {
+    async handle_post(req: IncomingMessage): Promise<number> {
         /**
          * Copy pasted from an issue on Deno's repository :)
          *
@@ -101,7 +101,7 @@ export class WhatsAppAPI extends WhatsAppAPIMiddleware {
      * @returns The challenge string to be sent to the client
      * @throws The error code
      */
-    handle_get(req: IncomingMessage) {
+    handle_get(req: IncomingMessage): string {
         try {
             return this.get(
                 Object.fromEntries(

--- a/src/middleware/sveltekit.ts
+++ b/src/middleware/sveltekit.ts
@@ -31,7 +31,7 @@ export class WhatsAppAPI extends WebStandardMiddleware {
      * @param req - The request object
      * @returns The status code to be sent to the client
      */
-    async handle_post(req: Request) {
+    async handle_post(req: Request): Promise<number> {
         return super.handle_post(req);
     }
 
@@ -63,7 +63,7 @@ export class WhatsAppAPI extends WebStandardMiddleware {
      * @returns The challenge string to be sent to the client
      * @throws The error code
      */
-    handle_get(req: Request) {
+    handle_get(req: Request): string {
         return super.handle_get(req);
     }
 }

--- a/src/middleware/vercel.ts
+++ b/src/middleware/vercel.ts
@@ -43,7 +43,7 @@ export class WhatsAppAPI extends NodeHTTPMiddleware {
      * @param req - The request object
      * @returns The status code to be sent to the client
      */
-    handle_post(req: VercelRequest) {
+    handle_post(req: VercelRequest): Promise<number> {
         return super.handle_post(req);
     }
 
@@ -78,7 +78,7 @@ export class WhatsAppAPI extends NodeHTTPMiddleware {
      * @returns The challenge string to be sent to the client
      * @throws The error code
      */
-    handle_get(req: VercelRequest) {
+    handle_get(req: VercelRequest): string {
         try {
             return this.get(req.query as GetParams);
         } catch (e) {

--- a/src/middleware/web-standard.ts
+++ b/src/middleware/web-standard.ts
@@ -14,7 +14,7 @@ export class WhatsAppAPI extends WhatsAppAPIMiddleware {
      * @param req - The request object
      * @returns The status code to be sent to the client
      */
-    async handle_post(req: Request) {
+    async handle_post(req: Request): Promise<number> {
         try {
             const body = await req.text();
 
@@ -39,7 +39,7 @@ export class WhatsAppAPI extends WhatsAppAPIMiddleware {
      * @returns The challenge string to be sent to the client
      * @throws The error code
      */
-    handle_get(req: Request) {
+    handle_get(req: Request): string {
         try {
             return this.get(
                 Object.fromEntries(new URL(req.url).searchParams) as GetParams

--- a/src/types.ts
+++ b/src/types.ts
@@ -790,9 +790,15 @@ export type ServerErrorResponse = {
     };
 };
 
-export type ServerSuccessResponse = {
-    success: true;
+export type NoServerError = {
+    error: never;
 };
+
+export type ServerSuccessResponse =
+    | {
+          success: true;
+      }
+    | NoServerError;
 
 export type ServerSentMessageResponse = {
     messaging_product: "whatsapp";
@@ -811,7 +817,7 @@ export type ServerSentMessageResponse = {
 };
 
 export type ServerMessageResponse =
-    | ServerSentMessageResponse
+    | (ServerSentMessageResponse | NoServerError)
     | ServerErrorResponse;
 
 export type ServerMarkAsReadResponse =
@@ -825,15 +831,22 @@ export type ServerQR = {
     qr_image_url?: string;
 };
 
-export type ServerCreateQRResponse = ServerQR | ServerErrorResponse;
-
-export type ServerRetrieveQRResponse =
-    | {
-          data: ServerQR[];
-      }
+export type ServerCreateQRResponse =
+    | (ServerQR | NoServerError)
     | ServerErrorResponse;
 
-export type ServerUpdateQRResponse = ServerQR | ServerErrorResponse;
+export type ServerRetrieveQRResponse =
+    | (
+          | {
+                data: ServerQR[];
+            }
+          | NoServerError
+      )
+    | ServerErrorResponse;
+
+export type ServerUpdateQRResponse =
+    | (ServerQR | NoServerError)
+    | ServerErrorResponse;
 
 export type ServerDeleteQRResponse =
     | ServerSuccessResponse
@@ -843,7 +856,9 @@ export type ServerMedia = {
     id: string;
 };
 
-export type ServerMediaUploadResponse = ServerMedia | ServerErrorResponse;
+export type ServerMediaUploadResponse =
+    | (ServerMedia | NoServerError)
+    | ServerErrorResponse;
 
 export type ValidMimeTypes =
     | "audio/aac"
@@ -866,13 +881,16 @@ export type ValidMimeTypes =
     | "image/webp";
 
 export type ServerMediaRetrieveResponse =
-    | ({
-          messaging_product: "whatsapp";
-          url: string;
-          mime_type: ValidMimeTypes;
-          sha256: string;
-          file_size: string;
-      } & ServerMedia)
+    | (
+          | ({
+                messaging_product: "whatsapp";
+                url: string;
+                mime_type: ValidMimeTypes;
+                sha256: string;
+                file_size: string;
+            } & ServerMedia)
+          | NoServerError
+      )
     | ServerErrorResponse;
 
 export type ServerMediaDeleteResponse =

--- a/src/types.ts
+++ b/src/types.ts
@@ -145,10 +145,14 @@ export type WhatsAppAPIConstructorArguments = TheBasicConstructorArguments &
 export abstract class ClientMessage {
     /**
      * The message type
+     *
+     * @internal
      */
     abstract get _type(): ClientMessageNames;
     /**
      * The message built as a string. In most cases it's just JSON.stringify(this)
+     *
+     * @internal
      */
     _build(): string {
         return JSON.stringify(this);
@@ -158,6 +162,8 @@ export abstract class ClientMessage {
 export interface ClientTypedMessageComponent {
     /**
      * The message's component type
+     *
+     * @internal
      */
     get _type(): string;
 }
@@ -165,6 +171,8 @@ export interface ClientTypedMessageComponent {
 export interface ClientBuildableMessageComponent {
     /**
      * The message's component builder method
+     *
+     * @internal
      */
     _build(..._: unknown[]): unknown;
 }
@@ -245,6 +253,7 @@ export abstract class ContactComponent
 {
     /**
      * @override
+     * @internal
      */
     _build(): unknown {
         return this;
@@ -252,8 +261,14 @@ export abstract class ContactComponent
 
     /**
      * Whether the component can be repeated multiple times in a contact.
+     *
+     * @internal
      */
     abstract get _many(): boolean;
+    /**
+     * @override
+     * @internal
+     */
     abstract get _type(): string;
 }
 
@@ -263,11 +278,16 @@ export abstract class ContactComponent
 export abstract class ContactMultipleComponent extends ContactComponent {
     /**
      * @override
+     * @internal
      */
     get _many(): true {
         return true;
     }
 
+    /**
+     * @override
+     * @internal
+     */
     abstract get _type(): string;
 }
 
@@ -277,11 +297,16 @@ export abstract class ContactMultipleComponent extends ContactComponent {
 export abstract class ContactUniqueComponent extends ContactComponent {
     /**
      * @override
+     * @internal
      */
     get _many(): false {
         return false;
     }
 
+    /**
+     * @override
+     * @internal
+     */
     abstract get _type(): string;
 }
 
@@ -291,6 +316,7 @@ export abstract class ContactUniqueComponent extends ContactComponent {
 export interface InteractiveAction extends ClientTypedMessageComponent {
     /**
      * @overload
+     * @internal
      */
     get _type():
         | "list"

